File: cache/cache-annotation-processor/src/test/java/ru/tinkoff/kora/cache/annotation/processor/CacheRunner.java
package ru.tinkoff.kora.cache.annotation.processor;

import jakarta.annotation.Nonnull;
import org.mockito.Mockito;
import ru.tinkoff.kora.cache.caffeine.$CaffeineCacheConfig_CaffeineTelemetryConfig_CaffeineLoggingConfig_ConfigValueExtractor;
import ru.tinkoff.kora.cache.caffeine.$CaffeineCacheConfig_CaffeineTelemetryConfig_CaffeineMetricsConfig_ConfigValueExtractor;
import ru.tinkoff.kora.cache.caffeine.CaffeineCacheConfig;
import ru.tinkoff.kora.cache.redis.RedisCacheClient;
import ru.tinkoff.kora.cache.redis.RedisCacheConfig;

import java.nio.ByteBuffer;
import java.time.Duration;
import java.util.*;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

import static org.mockito.Mockito.when;

final class CacheRunner {

    private CacheRunner() {}

    public static CaffeineCacheConfig getCaffeineConfig() {
        var config = Mockito.mock(CaffeineCacheConfig.class);
        var telemetry = Mockito.mock(CaffeineCacheConfig.CaffeineTelemetryConfig.class);
        when(config.telemetry()).thenReturn(telemetry);
        when(config.maximumSize()).thenReturn(100_000L);
        when(config.expireAfterAccess()).thenReturn(null);
        when(config.expireAfterWrite()).thenReturn(null);
        when(telemetry.metrics()).thenReturn(new $CaffeineCacheConfig_CaffeineTelemetryConfig_CaffeineMetricsConfig_ConfigValueExtractor.CaffeineMetricsConfig_Defaults());
        when(telemetry.logging()).thenReturn(new $CaffeineCacheConfig_CaffeineTelemetryConfig_CaffeineLoggingConfig_ConfigValueExtractor.CaffeineLoggingConfig_Defaults());
        return config;
    }

    public static RedisCacheConfig getRedisConfig() {
        var config = Mockito.mock(RedisCacheConfig.class);
        when(config.keyPrefix()).thenReturn("pref");
        return config;
    }

    public static RedisCacheClient lettuceClient(final Map<ByteBuffer, ByteBuffer> cache) {
        return new RedisCacheClient() {
            @Override
            public CompletionStage<List<byte[]>> scan(byte[] prefix) {
                List<byte[]> keys = new ArrayList<>();
                for (ByteBuffer buffer : cache.keySet()) {
                    if (Arrays.equals(Arrays.copyOf(buffer.array(), prefix.length), prefix)) {
                        keys.add(buffer.array());
                    }
                }
                return CompletableFuture.completedFuture(keys);
            }

            @Override
            public CompletionStage<byte[]> get(byte[] key) {
                var r = cache.get(ByteBuffer.wrap(key));
                return (r == null)
                    ? CompletableFuture.completedFuture(null)
                    : CompletableFuture.completedFuture(r.array());
            }

            @Nonnull
            @Override
            public CompletionStage<Map<byte[], byte[]>> mget(byte[][] keys) {
                final Map<byte[], byte[]> result = new HashMap<>();
                for (byte[] key : keys) {
                    Optional.ofNullable(cache.get(ByteBuffer.wrap(key))).ifPresent(r -> result.put(key, r.array()));
                }
                return CompletableFuture.completedFuture(result);
            }

            @Override
            public CompletionStage<byte[]> getex(byte[] key, long expireAfterMillis) {
                return get(key);
            }

            @Nonnull
            @Override
            public CompletionStage<Map<byte[], byte[]>> getex(byte[][] keys, long expireAfterMillis) {
                return mget(keys);
            }

            @Override
            public CompletionStage<Boolean> set(byte[] key, byte[] value) {
                cache.put(ByteBuffer.wrap(key), ByteBuffer.wrap(value));
                return CompletableFuture.completedFuture(true);
            }

            @Override
            public CompletionStage<Boolean> mset(Map<byte[], byte[]> keyAndValue) {
                keyAndValue.forEach((k, v) -> cache.put(ByteBuffer.wrap(k), ByteBuffer.wrap(v)));
                return CompletableFuture.completedFuture(true);
            }

            @Override
            public CompletionStage<Boolean> psetex(Map<byte[], byte[]> keyAndValue, long expireAfterMillis) {
                return mset(keyAndValue);
            }

            @Override
            public CompletionStage<Boolean> psetex(byte[] key, byte[] value, long expireAfterMillis) {
                return set(key, value);
            }

            @Override
            public CompletionStage<Long> del(byte[] key) {
                return CompletableFuture.completedFuture(cache.remove(ByteBuffer.wrap(key)) == null ? 0L : 1L);
            }

            @Override
            public CompletionStage<Long> del(byte[][] keys) {
                int counter = 0;
                for (byte[] key : keys) {
                    counter += del(key).toCompletableFuture().join();
                }
                return CompletableFuture.completedFuture((long) counter);
            }

            @Override
            public CompletionStage<Boolean> flushAll() {
                cache.clear();
                return CompletableFuture.completedFuture(true);
            }
        };
    }
}

File: cache/cache-symbol-processor/src/test/kotlin/ru/tinkoff/kora/cache/symbol/processor/CacheRunner.kt
package ru.tinkoff.kora.cache.symbol.processor

import kotlinx.coroutines.future.await
import kotlinx.coroutines.runBlocking
import org.mockito.Mockito
import org.mockito.Mockito.`when`
import ru.tinkoff.kora.cache.caffeine.`$CaffeineCacheConfig_CaffeineTelemetryConfig_CaffeineLoggingConfig_ConfigValueExtractor`
import ru.tinkoff.kora.cache.caffeine.`$CaffeineCacheConfig_CaffeineTelemetryConfig_CaffeineMetricsConfig_ConfigValueExtractor`
import ru.tinkoff.kora.cache.caffeine.CaffeineCacheConfig
import ru.tinkoff.kora.cache.caffeine.CaffeineCacheConfig.CaffeineTelemetryConfig
import ru.tinkoff.kora.cache.redis.RedisCacheClient
import ru.tinkoff.kora.cache.redis.RedisCacheConfig
import java.nio.ByteBuffer
import java.time.Duration
import java.util.*
import java.util.concurrent.CompletableFuture
import java.util.concurrent.CompletionStage


class CacheRunner {

    companion object {

        fun getCaffeineConfig(): CaffeineCacheConfig {
            val config = Mockito.mock(CaffeineCacheConfig::class.java)
            val telemetry = Mockito.mock(CaffeineTelemetryConfig::class.java)
            `when`(config.telemetry()).thenReturn(telemetry)
            `when`(config.maximumSize()).thenReturn(100000L)
            `when`(config.expireAfterAccess()).thenReturn(null)
            `when`(config.expireAfterWrite()).thenReturn(null)
            `when`(telemetry.metrics()).thenReturn(`$CaffeineCacheConfig_CaffeineTelemetryConfig_CaffeineMetricsConfig_ConfigValueExtractor`.CaffeineMetricsConfig_Defaults())
            `when`(telemetry.logging()).thenReturn(`$CaffeineCacheConfig_CaffeineTelemetryConfig_CaffeineLoggingConfig_ConfigValueExtractor`.CaffeineLoggingConfig_Defaults())
            return config
        }

        fun getRedisConfig(): RedisCacheConfig {
            return object : RedisCacheConfig {

                override fun keyPrefix(): String = "pref"

                override fun expireAfterWrite(): Duration? = null

                override fun expireAfterAccess(): Duration? = null
            }
        }

        fun lettuceClient(cache: MutableMap<ByteBuffer?, ByteBuffer?>): RedisCacheClient {
            return object : RedisCacheClient {

                override fun scan(prefix: ByteArray): CompletionStage<List<ByteArray>> {
                    val keys: MutableList<ByteArray> = ArrayList()
                    for (buffer in cache.keys) {
                        if (buffer != null) {
                            if (buffer.array().copyOf(prefix.size).contentEquals(prefix)) {
                                keys.add(buffer.array())
                            }
                        }
                    }
                    return CompletableFuture.completedFuture(keys)
                }

                override fun get(key: ByteArray): CompletionStage<ByteArray?> {
                    val r = cache[ByteBuffer.wrap(key)]
                    return CompletableFuture.completedFuture(r?.array())
                }

                override fun mget(keys: Array<ByteArray>): CompletionStage<Map<ByteArray, ByteArray>> {
                    val result: MutableMap<ByteArray, ByteArray> = HashMap()
                    for (key in keys) {
                        Optional.ofNullable(cache[ByteBuffer.wrap(key)]).ifPresent { r: ByteBuffer ->
                            result[key] = r.array()
                        }
                    }
                    return CompletableFuture.completedFuture(result)
                }

                override fun getex(key: ByteArray, expireAfterMillis: Long): CompletionStage<ByteArray?> {
                    return get(key)
                }

                override fun getex(keys: Array<ByteArray>, expireAfterMillis: Long): CompletionStage<Map<ByteArray, ByteArray>> {
                    return mget(keys)
                }

                override fun set(key: ByteArray, value: ByteArray) : CompletionStage<Boolean> {
                    cache[ByteBuffer.wrap(key)] = ByteBuffer.wrap(value)
                    return CompletableFuture.completedFuture(true)
                }

                override fun mset(keyAndValue: MutableMap<ByteArray, ByteArray>) : CompletionStage<Boolean> {
                    keyAndValue.forEach { (k, v) -> set(k, v) }
                    return CompletableFuture.completedFuture(true)
                }

                override fun psetex(keyAndValue: MutableMap<ByteArray, ByteArray>, expireAfterMillis: Long): CompletionStage<Boolean> {
                    mset(keyAndValue)
                    return CompletableFuture.completedFuture(true)
                }

                override fun psetex(key: ByteArray, value: ByteArray, expireAfterMillis: Long): CompletionStage<Boolean> {
                    return set(key, value)
                }

                override fun del(key: ByteArray): CompletionStage<Long> {
                    val res = if (cache.remove(ByteBuffer.wrap(key)) == null) 0L else 1L
                    return CompletableFuture.completedFuture(res)
                }

                override fun del(keys: Array<ByteArray>): CompletionStage<Long> {
                    var counter = 0
                    for (key in keys) {
                        counter += runBlocking { del(key).await().toInt() }
                    }
                    return CompletableFuture.completedFuture(counter.toLong())
                }

                override fun flushAll() : CompletionStage<Boolean> {
                    cache.clear()
                    return CompletableFuture.completedFuture(true)
                }
            }
        }
    }
}
